fetch_from_gmail.py

```diff
@@ -0,0 +1,374 @@
+import os.path
+import base64
+from googleapiclient.discovery import build
+from typing import List, Dict, Set
+import json
+import fcntl
+from concurrent.futures import ThreadPoolExecutor, as_completed
+import os
+import dotenv
+from base64 import urlsafe_b64decode
+from email_reply_parser import EmailReplyParser
+from bs4 import BeautifulSoup
+
+dotenv.load_dotenv()
+
+# If modifying scopes, delete token.json
+SCOPES = ["https://www.googleapis.com/auth/gmail.readonly"]
+EMAILS_BATCH_SIZE = int(os.getenv("EMAIL_FETCHING_BATCH_SIZE"))
+MAX_WORKERS = int(os.getenv("EMAIL_FETCHING_MAX_PARALLEL_WORKERS"))
+
+last_page_token = None
+
+
+class Email:
+    # date is in MS since epoch
+    def __init__(self, id: str, from_user: str, body: str, date: int, subject: str):
+        self.id = id
+        self.from_user = from_user
+        self.body = body
+        self.date = date
+        self.subject = subject
+
+    def to_dict(self):
+        return {
+            "id": self.id,
+            "body": self.body,
+            "date": self.date,
+            "from_user": self.from_user,
+            "subject": self.subject,
+        }
+
+    @classmethod
+    def from_dict(cls, data):
+        return cls(
+            id=data["id"],
+            body=data["body"],
+            date=data["date"],
+            from_user=data["from_user"],
+            subject=data["subject"],
+        )
+
+
+class EmailThread:
+    def __init__(self, thread_id: str, emails: List[Email]):
+        self.thread_id = thread_id
+        self.emails = sorted(emails, key=lambda x: x.date)
+
+    def add_email(self, email: Email):
+        # if the email is already in the thread, don't add it
+        if email.id in [e.id for e in self.emails]:
+            return
+        self.emails.append(email)
+        self.emails.sort(key=lambda x: x.date)
+
+    def to_dict(self):
+        return {
+            "thread_id": self.thread_id,
+            "emails": [email.to_dict() for email in self.emails],
+        }
+
+    @classmethod
+    def from_dict(cls, data):
+        emails = [Email.from_dict(email_data) for email_data in data["emails"]]
+        return cls(thread_id=data["thread_id"], emails=emails)
+
+
+threads_dict: Dict[str, EmailThread] = {}
+
+seen_email_ids: Set[str] = set()
+
+
+def _clean_plain(text: str) -> str:
+    """Return only the new part of a plain-text message."""
+    # library handles signatures & quoting reliably
+    return EmailReplyParser.parse_reply(text).strip()
+
+
+def _clean_html(html: str) -> str:
+    """Remove quoted history from a Gmail / generic HTML body."""
```

Comment:
Would this be a problem when some inlines their response?

Reply:
good catch! it was a problem. I have fixed that, and moved email body cleanup to another file. We do a few things to clean up the body:
- remove html tags
- remove threaded messages within a message body. Threaded messages start with something like: `"On Thu, Jul 10, 2025 at 1:53 AM Rishabh Poddar <rishabh.poddar@gmail.com>\nwrote"`, but vary a lot (sometimes there are extra `\n` in the middle, or there is a space after the `<` followed by the email etc). So as soon as this is detected, it discards everything below that part from the email